File: tg_bot_python_with_reqs/tg.py
from flask import Flask, request
from config import Config
from functions import send_answer, get_requisites_epay, log_error
import time
from pathlib import Path

app = Flask(__name__)
BASE_DIR = Path(__file__).resolve().parent

@app.route("/action/tg", methods=["POST"])
def tg_webhook():
    try:
        update = request.get_json(force=True)
        chat_id = update["message"]["chat"]["id"]
        message = update["message"].get("text", "")

        if message == "/start":
            send_answer(Config.text_before_enter_price, chat_id)
            return "ok", 200

        data_for_api = {
            "amount": message,
            "merchant_order_id": "optional",
            "api_key": Config.api_secret,
            "notice_url": f"{request.url_root}action/callback"
        }

        send_answer("⌛Ожидаем реквизиты...", chat_id)
        result_order_data = get_requisites_epay(data_for_api)

        if result_order_data.get("error_desc"):
            send_answer("Ошибка! " + result_order_data["error_desc"], chat_id, False)
        else:
            order_data = [
                time.strftime("%Y-%m-%d %H:%M:%S"),
                str(result_order_data["order_id"]),
                str(chat_id),
                str(data_for_api["amount"])
            ]
            with open(BASE_DIR / "orders.txt", "a", encoding="utf-8") as f:
                f.write("|".join(order_data) + "\n")

            receiver_wallet_str = f" *Номер карты для оплаты*: `{result_order_data.get('card_number','')}`"

            if (result_order_data.get("qr_sbp_url") or result_order_data.get("card_form_url")) and not result_order_data.get("card_number"):
                url = result_order_data.get("qr_sbp_url") or result_order_data.get("card_form_url")
                receiver_wallet_str = f" *Ссылка на оплату*: [{url}]({url}) "

            msg = (
                f"{Config.code_symbol_order} *Создан заказ*: №{result_order_data['order_id']}\n\n"
                f"{Config.code_symbol_card}{receiver_wallet_str} \n"
                f"{Config.code_symbol_amount} *Сумма платежа*: `{result_order_data['amount']}` {Config.code_symbol_rub}\n\n"
                f"{Config.code_symbol_timer} *Время на оплату*: {Config.time_minutes_for_payment} мин"
            )
            send_answer(msg, chat_id)

            if Config.text_after_requisites:
                send_answer(Config.text_after_requisites, chat_id)

        return "ok", 200
    except Exception as e:
        log_error(str(e))
        return "error", 200

if __name__ == "__main__":
    app.run(host="0.0.0.0", port=5000)

File: Test10921820_bot (1)/action/tg.php
<?php

require_once $_SERVER['DOCUMENT_ROOT']."/config.php";
require_once $_SERVER['DOCUMENT_ROOT']."/functions.php";

$update = file_get_contents('php://input');

$update = json_decode($update,TRUE);

$chatId = $update["message"]["chat"]["id"];

$message = $update["message"]["text"];

if ($message === '/start') {
    send_answer(Config::$textBeforeEnterPrice,$chatId);
    header("HTTP/1.1 200 OK");
    exit;
} else {
    $dataForApi = [
        'amount'            => $message,
        'merchant_order_id' => 'optional',
        'api_key'           => Config::$apiSecret,
        'notice_url'        => "https://".$_SERVER['SERVER_NAME']."/action/callback.php",
    ];

    try {
        send_answer("\xE2\x8C\x9B"."Ожидаем реквизиты...", $chatId);

        $resultOrderData = json_decode(getRequisitesEpay($dataForApi),true);

        if (!empty($resultOrderData['error_desc'])) {
            send_answer("Ошибка! " . $resultOrderData['error_desc'], $chatId,false);
        } else {
            $orderData = [date("Y-m-d H:i:s"), $resultOrderData['order_id'], $chatId, $dataForApi['amount']];
            file_put_contents($_SERVER['DOCUMENT_ROOT'].'/orders.txt', implode('|', $orderData). "\r\n", FILE_APPEND);

            $receiverWalletStr = " *Номер карты для оплаты*: `".$resultOrderData['card_number']."`";

            if ((!empty($resultOrderData['qr_sbp_url']) || !empty($resultOrderData['card_form_url'])) && empty($resultOrderData['card_number'])) {
                $url = !empty($resultOrderData['qr_sbp_url']) ? $resultOrderData['qr_sbp_url'] : $resultOrderData['card_form_url'];

                $receiverWalletStr = " *Ссылка на оплату*: [".$url."](".$url.") ";
            }

            send_answer(Config::$codeSymbolOrder." *Создан заказ*: №".$resultOrderData['order_id']."\n\n".Config::$codeSymbolCard.$receiverWalletStr." \n".Config::$codeSymbolAmount." *Сумма платежа*: `" . $resultOrderData['amount']. "` ".Config::$codeSymbolRUB."\n\n".Config::$codeSymbolTimer." *Время на оплату*: ".Config::$timeMinutesForPayment." мин", $chatId);

            if (!empty(Config::$textAfterRequisites)) {
                send_answer(Config::$textAfterRequisites, $chatId);
            }
        }

        header("HTTP/1.1 200 OK");
        exit;
    } catch (Exception $exception) {
        file_put_contents($_SERVER['DOCUMENT_ROOT'].'/response_tg_errors.txt', "DATE: ". date("Y-m-d H:i:s")." RESPONSE: ".$exception->getMessage(). "\r\n", FILE_APPEND);
    }
}

header("HTTP/1.1 200 OK");
exit;
